src/startup.js

```diff
@@ -1,132 +1,149 @@
+			var optionName =
+				camelize( attr.nodeName.indexOf("data-") === 0 ?
+					attr.nodeName.replace("data-","") :
+					attr.nodeName );
+			options[optionName] = (attr.value === "") ? true : attr.value;
+		});
 
-		// first set the config that is set with a steal object
-		if(config){
-			System.config(config);
+		// main source within steals script is deprecated
+		// and will be removed in future releases
+		var source = script.innerHTML;
+		if(/\S/.test(source)){
+			options.mainSource = source;
 		}
 
-		// B: DO THINGS WITH OPTIONS
-		// CALCULATE CURRENT LOCATION OF THINGS ...
-		System.config(urlOptions);
 
+		return options;
+	};
 
-		setEnvsConfig.call(this.System);
+	var getUrlOptions = function (){
+		return new Promise(function(resolve, reject){
 
-		// Read the env now because we can't overwrite everything yet
+			// Get options from the script tag
+			if (isWebWorker) {
+				resolve({
+					stealURL: location.href
+				});
 
-		// immediate steals we do
-		var steals = [];
+			} else if(isBrowserWithWindow) {
+				// if the browser supports currentScript, us it!
+				if (document.currentScript) {
+					resolve(getScriptOptions(document.currentScript));
+					return;
+				}
 
-		// we only load things with force = true
-		if ( System.loadBundles ) {
+				// dealing with async & deferred scripts
+				// set an onload handler for all script tags and the first one which executes
+				// is your stealjs
+				var scripts = document.scripts;
+				function onLoad() {
+					for (var i = 0; i < scripts.length; ++i) {
+						scripts[i].removeEventListener('load', onLoad, false);
+					}
+					resolve(getScriptOptions(event.target));
+				}
+				for (var i = 0; i < scripts.length; ++i) {
+					scripts[i].addEventListener('load', onLoad, false);
+				}
 
-			if(!System.main && System.isEnv("production") && !System.stealBundled) {
-				// prevent this warning from being removed by Uglify
-				var warn = console && console.warn || function() {};
-				warn.call(console, "Attribute 'main' is required in production environment. Please add it to the script tag.");
+			} else {
+				// or the only option is where steal is.
+				resolve({
+					stealPath: __dirname
+				});
 			}
+		})
+	};
 
-			configDeferred = System["import"](System.configMain);
+	steal.startup = function(config){
+		appPromise = getUrlOptions().then(function(urlOptions) {
 
-			appDeferred = configDeferred.then(function(cfg){
-				setEnvsConfig.call(System);
-				return System.main ? System["import"](System.main) : cfg;
-			});
+			if (typeof config === 'object') {
+				// the url options are the source of truth
+				config = extend(config, urlOptions);
+			} else {
+				config = urlOptions;
+			}
 
-		} else {
-			configDeferred = System["import"](System.configMain);
+			// set the config
+			System.config(config);
 
-			devDeferred = configDeferred.then(function(){
-				setEnvsConfig.call(System);
-				setupLiveReload.call(System);
+			setEnvsConfig.call(this.System);
```

Comment:
`this` is no longer going to be `steal` since it's inside of a Promise then function.  Can put `var steal = this` before the `getUrlOptions()` call and refer to that.